packages/app/src/components/inspector/PatternInspector.tsx

```diff
@@ -0,0 +1,242 @@
+    cursor: 'pointer',
+    userSelect: 'auto',
+    alignSelf: 'center',
+    height: '30px',
+    width: '105px',
+    borderRadius: '24px',
+    border: 'none',
+    display: show ? 'block' : 'none',
+    zIndex: 1,
+  }),
+);
+
+interface Props {
+  id: string;
+  images: FileMap;
+  pattern: SketchPattern;
+  onChangeImage?: (image: Sketch.FileRef | Sketch.DataRef) => void;
+  onChangeFillType?: (amount: Sketch.PatternFillType) => void;
+  onChangeTileScale?: (amount: number) => void;
+  createImage: (image: ArrayBuffer, _ref: string) => void;
+}
+
+export type PatternFillTypes = 'Stretch' | 'Fill' | 'Fit' | 'Tile';
+export const patternFillTypeOptions: PatternFillTypes[] = [
+  'Tile',
+  'Fill',
+  'Stretch',
+  'Fit',
+];
+
+export default memo(function PatternInspector({
+  id,
+  images,
+  pattern,
+  createImage,
+  onChangeImage,
+  onChangeFillType,
+  onChangeTileScale,
+}: Props) {
+  const [isHovering, onHoverChange] = useState(false);
+  const { hoverProps } = useHover({
+    onHoverChange,
+  });
+
+  const patternType = pattern.patternFillType;
+  const isTile = patternType === Sketch.PatternFillType.Tile;
+
+  const changeFillType = useCallback(
+    (value: PatternFillTypes) => {
+      onChangeFillType?.(Sketch.PatternFillType[value]);
+    },
+    [onChangeFillType],
+  );
+
+  const onSubmitTileScale = useCallback(
+    (value: number) => {
+      onChangeTileScale?.(value / 100);
+    },
+    [onChangeTileScale],
+  );
+
+  const onNudgeTileScale = useCallback(
+    (value: number) => {
+      onChangeTileScale?.(value / 100);
+    },
+    [onChangeTileScale],
+  );
+
+  const background = useMemo(
+    () => getPatternBackground(images, pattern.image),
+    [images, pattern.image],
+  );
+
+  const backgroundSize = useMemo(
+    () => getPatternSize(pattern.patternFillType, pattern.patternTileScale),
+    [pattern.patternFillType, pattern.patternTileScale],
+  );
+
+  const handleImageFile = useCallback(
+    async (file: File) => {
+      const extension = SUPPORTED_FILE_TYPES[file.type];
+
+      if (!extension) {
+        alert(
+          `Files of type ${file.type} aren't supported. Files of type png, jpg, and webp are supported.`,
+        );
+        return;
+      }
+
+      const data = await file.arrayBuffer();
+      const _ref = `images/${uuid()}.${extension}`;
+
+      createImage(data, _ref);
+      onChangeImage?.({
+        _class: 'MSJSONFileReference',
+        _ref: _ref,
+        _ref_class: 'MSImageData',
+      });
+    },
+    [createImage, onChangeImage],
+  );
+
+  const openFile = useCallback(async () => {
+    const file = await fileOpen({
+      extensions: Object.values(SUPPORTED_FILE_TYPES).map((e) => '.' + e),
+      mimeTypes: Object.keys(SUPPORTED_FILE_TYPES),
+    });
+
+    handleImageFile(file);
+  }, [handleImageFile]);
+
+  const handleDropEvent = useCallback(
+    async (e: DragEvent) => {
+      e.preventDefault();
+      for (const file of e.dataTransfer.items) {
```

Comment:
This looks like it will call `handleImageFile` for each file in a list, but we probably only want to handle one. We could do:

```ts
const file = e.dataTransfer.items[0]

if (!file) return 
      
const entry = ...;
```

We should probably also show some user feedback with `alert` if the `kind` is wrong or we can't get the file as an `entry`

Reply:
What should it say when we have the wrong kind or entry?

Reply:
- For wrong kind, "Only files are supported"
- For invalid entry, "Unable to read file ${filename}"

Reply:
We don't use kind anymore, I change the to files instead of items. Because items it's used mostly when you want to also allow to drop directories and I don't think we are interested in that.

Reply:
I can't get the file name if the file it's invalid 😅